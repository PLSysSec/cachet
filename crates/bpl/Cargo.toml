[package]
name = "bpl"
version = "0.0.0"
authors = ["Michael Smith <michael@spinda.net>"]
license = "MPL-2.0"
description = """
AST, parser, and printer for the Boogie intermediate verification language
"""
edition = "2021"

[build-dependencies]
anyhow = "1.0.28"
lalrpop = { git = "https://github.com/PLSysSec/lalrpop", branch = "cachet" }
proc-macro2 = "1.0.49"
quote = "1.0.23"
walkdir = "2.3.2"

[dependencies]
anyhow = "1.0.28"
codespan = "0.11.1"
codespan-reporting = "0.11.1"
derive_more = "0.99.16"
indent_write = "2.2.0"
internment = "0.5.4"
joinery = "3.1.0"
lalrpop-util = { git = "https://github.com/PLSysSec/lalrpop", branch = "cachet" }
lazy_static = "1.4.0"
phf = { version = "0.10.1", features = ["macros"] }

[dev-dependencies]
insta = "1.23.0"
regex = "1.7.0"
similar-asserts = "1.4.2"
