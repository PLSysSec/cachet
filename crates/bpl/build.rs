use std::env;
use std::fs::File;
use std::io::prelude::*;
use std::path::{Component, Path, PathBuf};

use anyhow::{Context, Error};
use proc_macro2::{Ident, TokenStream};
use quote::{format_ident, quote};
use walkdir::WalkDir;

fn main() -> Result<(), Error> {
    lalrpop::Configuration::new()
        .emit_rerun_directives(true)
        .process_current_dir()
        .expect("lalrpop error");

    let mut test_cases = TestCases::default();
    test_cases.scan_dir(PathBuf::from("tests"), "samples")?;
    test_cases.scan_dir(PathBuf::from_iter(["..", "..", "vendor"]), "boogie")?;
    let test_src = test_cases.generate_src();

    let test_file_path =
        Path::new(&env::var("OUT_DIR").context("Couldn't read `OUT_DIR` environment variable")?)
            .join("test.rs");
    let mut test_file = File::create(&test_file_path).context("Couldn't create `test.rs`")?;
    write!(test_file, "{}", test_src).context("Couldn't write `test.rs`")?;

    Ok(())
}

#[derive(Default)]
struct TestCases {
    fn_idents: Vec<Ident>,
    names: Vec<String>,
    paths: Vec<String>,
}

impl TestCases {
    fn scan_dir(&mut self, base_dir_path: PathBuf, scan_dir_name: &str) -> Result<(), Error> {
        let scan_dir_path = base_dir_path.join(scan_dir_name);
        println!("cargo:rerun-if-changed={}", scan_dir_path.display());

        for entry in WalkDir::new(&scan_dir_path) {
            let entry = entry
                .with_context(|| format!("Error walking directory {}", scan_dir_path.display()))?;

            if !entry.file_type().is_file() {
                continue;
            }

            let path = entry.path();

            let extension = match path.extension() {
                Some(extension) => extension,
                None => continue,
            };
            if extension != "bpl" {
                continue;
            }

            let stripped_path = path
                .strip_prefix(&base_dir_path)
                .context("Bad path prefix")?
                .with_extension("");

            let mut name = String::new();
            for component in stripped_path.components() {
                if let Component::Normal(component) = component {
                    if !name.is_empty() {
                        name.push_str("__");
                    }
                    name.extend(
                        component
                            .to_string_lossy()
                            .chars()
                            .map(|c| if c.is_alphanumeric() { c } else { '_' }),
                    );
                }
            }

            self.fn_idents.push(format_ident!("test__{}", name));
            self.names.push(name);

            // TODO(spinda): Is there a better way to do this than
            // `to_string_lossy`?
            self.paths.push(path.to_string_lossy().into_owned());
        }

        Ok(())
    }
    
    fn generate_src(&self) -> TokenStream {
        let TestCases { fn_idents, names, paths } = self;
        quote! {
            #(
                #[test]
                #[allow(non_snake_case)]
                fn #fn_idents() -> Result<(), Error> {
                    test_program(#names, #paths)
                }
            )*
        }
    }
}
