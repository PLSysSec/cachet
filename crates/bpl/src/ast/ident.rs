use std::borrow::Borrow;
use std::fmt::{self, Debug};
use std::ops::Deref;

use derive_more::Display;
use internment::Intern;

/// [Grammar](https://boogie-docs.readthedocs.io/en/latest/LangRef.html#grammar-token-ident):
///
/// ```grammar
/// ident     ::= [ "\\" ] non_digit { non_digit | digit }
/// non_digit ::= ( "A…Z" | "a…z" | "'" | "~" | "#" | "$" | "^" | "_" | "." | "?" | "`" )
/// digit     ::= "0…9"
/// ```
// TODO(spinda): Expand derives.
#[derive(Clone, Copy, Display, Eq, Hash, Ord, PartialEq, PartialOrd)]
pub struct Ident(Intern<String>);

impl<T: ?Sized> AsRef<T> for Ident
where
    String: AsRef<T>,
{
    fn as_ref(&self) -> &T {
        (**self).as_ref()
    }
}

impl<T: ?Sized> Borrow<T> for Ident
where
    String: Borrow<T>,
{
    fn borrow(&self) -> &T {
        (**self).borrow()
    }
}

impl Deref for Ident {
    type Target = String;

    fn deref(&self) -> &Self::Target {
        &*self.0
    }
}

impl Debug for Ident {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        Debug::fmt(&**self, f)
    }
}

impl From<String> for Ident {
    fn from(s: String) -> Self {
        Ident(s.into())
    }
}

impl From<&String> for Ident {
    fn from(s: &String) -> Self {
        String::from(s).into()
    }
}

impl From<&str> for Ident {
    fn from(s: &str) -> Self {
        String::from(s).into()
    }
}

impl From<&mut str> for Ident {
    fn from(s: &mut str) -> Self {
        String::from(s).into()
    }
}

impl From<Ident> for String {
    fn from(ident: Ident) -> Self {
        (*ident).clone()
    }
}
