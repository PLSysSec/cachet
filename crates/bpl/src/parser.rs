// vim: set tw=99 ts=4 sts=4 sw=4 et:

mod error;
mod helpers;

use lalrpop_util::lalrpop_mod;
lalrpop_mod!(grammar, "/parser/grammar.rs");
use lazy_static::lazy_static;

use crate::ast::BoogieProgram;

pub use crate::parser::error::ParseError;

// * Parsing Entry Point

pub fn parse_boogie_program(src: &str) -> Result<BoogieProgram, ParseError> {
    BOOGIE_PROGRAM_PARSER.parse(src).map_err(ParseError::new)
}

lazy_static! {
    static ref BOOGIE_PROGRAM_PARSER: grammar::BoogieProgramParser =
        grammar::BoogieProgramParser::new();
}
