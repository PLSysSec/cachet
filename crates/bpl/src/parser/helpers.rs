// vim: set tw=99 ts=4 sts=4 sw=4 et:

use crate::ast::{BvLit, Dec, Exp, Float, FloatValue, PlainFloatValue, Spanned};

pub type RawParseError<T> = lalrpop_util::ParseError<usize, T, UserParseError>;
pub type UserParseError = Spanned<&'static str>;

pub fn decode_dec(input: &str) -> Dec {
    let (whole, fract, exp) = match input.split_once('.') {
        Some((whole, rest)) => {
            let (fract, exp) = decode_exp(rest);
            (whole, Some(fract), exp)
        }
        None => {
            let (whole, exp) = decode_exp(input);
            (whole, None, exp)
        }
    };
    Dec {
        whole: whole.to_owned(),
        fract: fract.map(ToOwned::to_owned),
        exp,
    }
}

fn decode_exp(input: &str) -> (&str, Option<Exp>) {
    match input.split_once('e') {
        Some((init, exp)) => {
            let (is_neg, digits) = take_neg(exp);
            (
                init,
                Some(Exp {
                    is_neg,
                    digits: digits.to_owned(),
                }),
            )
        }
        None => (input, None),
    }
}

pub fn decode_float(input: &str) -> Float {
    // Consume leading `-`.
    let (is_neg, rest) = match input.strip_prefix('-') {
        Some(rest) => (true, rest),
        None => (false, input),
    };

    // Skip `0`.
    let rest = &rest[1..];

    let (value, rest) = match rest.chars().next().expect("unexpected end of float") {
        'x' => {
            let rest = &rest[1..];
            let (value_input, rest) = rest
                .rsplit_once('f')
                .expect("missing float significand size");
            let (value_rest, exp) = value_input
                .rsplit_once('e')
                .expect("missing float exponent");
            let (exp_is_neg, exp_digits) = take_neg(exp);
            let (whole, fract) = value_rest
                .split_once('.')
                .expect("missing float fractional part");
            let value = PlainFloatValue {
                is_neg,
                whole: whole.to_owned(),
                fract: fract.to_owned(),
                exp: Exp {
                    is_neg: exp_is_neg,
                    digits: exp_digits.to_owned(),
                },
            };
            (value.into(), rest)
        }
        // `NaN...`
        'N' | 'n' => (FloatValue::NotANumber, &rest[3..]),
        // `+oo...`
        '+' => (FloatValue::PosInfinity, &rest[3..]),
        // `-oo...`
        '-' => (FloatValue::NegInfinity, &rest[3..]),
        _ => panic!("malformed float"),
    };

    let (sig_size, exp_size) = rest.split_once('e').expect("missing float exponent size");

    Float {
        value,
        sig_size: sig_size.to_owned(),
        exp_size: exp_size.to_owned(),
    }
}

pub fn decode_bv_lit(input: &str) -> BvLit {
    let (n, width) = input.split_once("bv").expect("malformed bitvector literal");
    BvLit {
        n: n.to_owned(),
        width: width.to_owned(),
    }
}

fn take_neg(input: &str) -> (bool, &str) {
    match input.strip_prefix('-') {
        Some(rest) => (true, rest),
        None => (false, input),
    }
}
