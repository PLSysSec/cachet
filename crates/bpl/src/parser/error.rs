// vim: set tw=99 ts=4 sts=4 sw=4 et:

use std::error::Error;
use std::fmt::{self, Display};

use codespan::{RawIndex, Span};
use codespan_reporting::diagnostic::{Diagnostic, Label, LabelStyle};
use phf::phf_map;

use crate::parser::helpers::RawParseError;

#[derive(Clone, Debug)]
pub struct ParseError {
    error: RawParseError<String>,
}

impl ParseError {
    pub(super) fn new<T: ToString>(error: RawParseError<T>) -> Self {
        Self {
            error: error.map_token(|token| token.to_string()),
        }
    }

    pub fn span(&self) -> Span {
        match &self.error {
            RawParseError::InvalidToken { location } => {
                Span::new(*location as RawIndex, *location as RawIndex)
            }
            RawParseError::UnrecognizedEOF { location, .. } => {
                Span::new(*location as RawIndex, *location as RawIndex)
            }
            RawParseError::UnrecognizedToken {
                token: (start, _, end),
                ..
            } => Span::new(*start as RawIndex, *end as RawIndex),
            RawParseError::ExtraToken {
                token: (start, _, end),
            } => Span::new(*start as RawIndex, *end as RawIndex),
            RawParseError::User { error } => return error.span,
        }
    }

    pub fn build_diagnostic(&self) -> Diagnostic<()> {
        self.build_diagnostic_for_file(())
    }

    pub fn build_diagnostic_for_file<T>(&self, file_id: T) -> Diagnostic<T> {
        Diagnostic::error()
            .with_message(self.to_string())
            .with_labels(vec![Label::new(LabelStyle::Primary, file_id, self.span())])
    }
}

impl Display for ParseError {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        match &self.error {
            RawParseError::InvalidToken { .. } => write!(f, "invalid token")?,
            RawParseError::UnrecognizedEOF { expected, .. } => {
                write!(f, "unexpected end of input")?;
                fmt_expected(f, expected)?;
            }
            RawParseError::UnrecognizedToken {
                token: (_, token, _),
                expected,
                ..
            } => {
                write!(f, "unexpected token `{}`", token)?;
                fmt_expected(f, expected)?;
            }
            RawParseError::ExtraToken {
                token: (_, token, _),
                ..
            } => {
                write!(f, "extra token `{}`", token)?;
            }
            RawParseError::User { error } => write!(f, "{}", error.value)?,
        }
        Ok(())
    }
}

impl Error for ParseError {}

static TOKEN_NAMES: phf::Map<&'static str, &'static str> = phf_map! {
    "r#\"\\\\\\\\?[A-Za-z'~#$^_.?`][A-Za-z0-9'~#$^_.?`]*\"#" => "identifier",
    "r#\"[0-9]+\"#" => "digits",
    "r#\"[0-9]+bv[0-9]+\"#" => "bitvector literal",
    "r#\"[0-9]+(e-?[0-9]+|\\\\.[0-9]+(e-?[0-9]+)?)\"#" => "decimal",
    "r#\"(-?0x[0-9A-Fa-f]+\\\\.[0-9A-Fa-f]+e-?[0-9]+f|0(NaN|nan|[+-]oo))[0-9]+e[0-9]+\"#" =>
        "float",
};

fn fmt_expected(f: &mut fmt::Formatter, expected: &[String]) -> Result<(), fmt::Error> {
    if !expected.is_empty() {
        write!(f, "; expected ")?;

        let mut tokens = expected.iter().enumerate().peekable();
        while let Some((token_index, token)) = tokens.next() {
            if token_index > 0 {
                if tokens.peek().is_some() {
                    write!(f, ", ")
                } else if token_index > 1 {
                    write!(f, ", or ")
                } else {
                    write!(f, " or ")
                }?;
            }

            match TOKEN_NAMES.get(token.as_str()) {
                Some(token_name) => write!(f, "{}", token_name),
                None => write!(f, "{}", token.replace("\"", "`")),
            }?;
        }
    }
    Ok(())
}
