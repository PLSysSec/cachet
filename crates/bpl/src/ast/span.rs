use std::fmt::{self, Debug, Display};
use std::error;

pub use codespan::{ByteIndex, RawIndex, Span};

#[derive(Clone, Copy)]
pub struct Spanned<T> {
    pub span: Span,
    pub value: T,
}

impl<T> Spanned<T> {
    pub const fn new(span: Span, value: T) -> Self {
        Self { span, value }
    }

    pub fn map<U>(self, f: impl FnOnce(T) -> U) -> Spanned<U> {
        Spanned {
            span: self.span,
            value: f(self.value),
        }
    }

    pub const fn as_ref(&self) -> Spanned<&T> {
        Spanned {
            span: self.span,
            value: &self.value,
        }
    }

    pub fn as_mut(&mut self) -> Spanned<&mut T> {
        Spanned {
            span: self.span,
            value: &mut self.value,
        }
    }
}

impl<T: Clone> Spanned<&T> {
    pub fn cloned(self) -> Spanned<T> {
        self.map(|value| value.clone())
    }
}

impl<T: Copy> Spanned<&T> {
    pub fn copied(self) -> Spanned<T> {
        self.map(|value| *value)
    }
}

impl<T: Debug> Debug for Spanned<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        Debug::fmt(&self.value, f)?;
        write!(f, " @ {}", self.span)?;
        Ok(())
    }
}

impl<T: Display> Display for Spanned<T> {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        Display::fmt(&self.value, f)
    }
}

impl<T: error::Error> error::Error for Spanned<T> {
    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
        self.value.source()
    }

    fn description(&self) -> &str {
        #[allow(deprecated)]
        self.value.description()
    }

    fn cause(&self) -> Option<&dyn error::Error> {
        #[allow(deprecated)]
        self.value.cause()
    }
}
