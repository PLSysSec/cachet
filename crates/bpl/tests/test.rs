// vim: set tw=99 ts=4 sts=4 sw=4 et:

use std::fmt::{self, Display};
use std::fs;
use std::io::Cursor;
use std::path::{Component, PathBuf};
use std::str;

use anyhow::Error;
use codespan_reporting::files::SimpleFile;
use codespan_reporting::term;
use codespan_reporting::term::termcolor::NoColor;
use insta::assert_snapshot;
use joinery::JoinableIterator;
use lazy_static::lazy_static;
use regex::Regex;
use similar_asserts::assert_eq;

use bpl::ast::BoogieProgram;
use bpl::parser::{parse_boogie_program, ParseError};

include!(concat!(env!("OUT_DIR"), "/test.rs"));

fn test_program(name: &str, path: &str) -> Result<(), Error> {
    let src = fs::read_to_string(path)?;
    let trimmed_src = src.trim();

    let (report, printed_program) = try_parse_and_print(path, &src)?;
    println!("{report}\n\n[Input - {path}]\n{}", trimmed_src);
    if let Some(printed_program) = printed_program {
        let (new_report, _) = try_parse_and_print(path, &printed_program)?;
        assert_eq!(report, new_report);
    }

    // For the snapshot, when printing the path, make sure, e.g., directory
    // separators get printed consistently between Linux and Windows.
    let snapshot_report = format!(
        "{report}\n\n[Input - {}]\n{}",
        OsIndependentPath(path),
        trimmed_src
    );
    assert_snapshot!(name, snapshot_report);
    Ok(())
}

fn try_parse_and_print(path: &str, src: &str) -> Result<(String, Option<String>), Error> {
    let parsed_program = match parse_boogie_program(&src) {
        Ok(parsed_program) => parsed_program,
        Err(parse_error) => {
            let report = format_parse_error(path, &src, &parse_error)?;
            return Ok((report, None));
        }
    };

    let printed_program = format!("{parsed_program}");
    let report = format!(
        "[Parsed]\n{}\n\n[Printed]\n{printed_program}",
        format_program(&parsed_program)
    );
    Ok((report, Some(printed_program)))
}

fn format_parse_error(path: &str, src: &str, parse_error: &ParseError) -> Result<String, Error> {
    let file = SimpleFile::new(path, src);
    let config = term::Config::default();
    let diagnostic = parse_error.build_diagnostic();

    let mut buf = Cursor::new(Vec::<u8>::new());
    term::emit(&mut NoColor::new(&mut buf), &config, &file, &diagnostic)?;
    let buf = buf.into_inner();

    let parse_error_str = format!("[Parse Error]\n{}", str::from_utf8(&buf)?.trim());
    Ok(parse_error_str.to_owned())
}

fn format_program(program: &BoogieProgram) -> String {
    let program_str = format!("{program:#?}");
    lazy_static! {
        static ref SPANNED_RE: Regex = Regex::new(r" @ \[[0-9]+, [0-9]+\)").unwrap();
    }
    SPANNED_RE.replace_all(&program_str, "").into_owned()
}

struct OsIndependentPath<'a>(&'a str);

impl Display for OsIndependentPath<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        write!(
            f,
            "{}",
            PathBuf::from(&self.0)
                .components()
                .map(OsIndependentComponent)
                .join_with("/")
        )
    }
}

struct OsIndependentComponent<'a>(Component<'a>);

impl Display for OsIndependentComponent<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
        match &self.0 {
            Component::Prefix(prefix) => write!(f, "{}", prefix.as_os_str().to_string_lossy())?,
            Component::RootDir => (),
            Component::CurDir => write!(f, ".")?,
            Component::ParentDir => write!(f, "..")?,
            Component::Normal(segment) => write!(f, "{}", segment.to_string_lossy())?,
        }
        Ok(())
    }
}
