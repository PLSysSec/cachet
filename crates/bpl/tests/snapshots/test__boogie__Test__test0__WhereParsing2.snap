---
source: crates/bpl/tests/test.rs
assertion_line: 41
expression: snapshot_report
---
[Parse Error]
error: unexpected token `where`; expected `;` or `extends`
  ┌─ ../../vendor/boogie/Test/test0/WhereParsing2.bpl:3:14
  │
3 │ const x: int where x < 0;  // error: where clauses not allowed on constants
  │              ^^^^^

[Input - ../../vendor/boogie/Test/test0/WhereParsing2.bpl]
// RUN: %boogie -noVerify "%s" > "%t"
// RUN: %diff "%s.expect" "%t"
const x: int where x < 0;  // error: where clauses not allowed on constants
